### DISCUSSION TITLE
problems about buliding lammps with mace and kokkos

### DISCUSSION THREAD

**Discussion by liubaoshuai1402** [ACEsuit/mace#673] [Q&A] [answered]
Dear developer,

I am trying to build lammps with kokkos and mace under the conduction of mace documentation.
Here are some problems I am countering.
**(1) I can not find the package ML-MACE in the lammps manual.**
But I find ML-PACE, Atomic Cluster Expansion potential, I want to know whether it's that lammps not updates its manual but in fact ML-MACE  has been available or ML-PACE is just the ML-MACE
<img width="980" alt="pace" src="https://github.com/user-attachments/assets/cd538c55-b141-4cd3-a726-ba249dd557e1">
(2) wrong cmake parameters about kokkos package in MACE example?
accourding to MACE's user guide at here: https://mace-docs.readthedocs.io/en/latest/guide/lammps.html#instructions-for-gpu
I find the example use the cmake keywords "-D Kokkos_ARCH_AMPERE100=ON"
**But I have visited the web of kokkos and no find the support for  AMPERE100.** only AMPERE 80 and AMPERE 86 is available!
**So I guess it's wrong that the example of MACE** ,"-D Kokkos_ARCH_AMPERE80=ON" will make kokkos support NVIDIA A100 GPU. 
Because I use the supercomputers  from China, NVIDIA A800 GPU is available, a weaker-version of NVIDIA A100 GPU.
**How should I set this cmake parameter?** I guess "-D Kokkos_ARCH_AMPERE80=ON" is ok?
<img width="1280" alt="kokkos" src="https://github.com/user-attachments/assets/30df37ae-6d40-4d18-a2e5-b26da98b955c">

**Comment by wcwitt** [comment, marked as answer]
(1) The MACE pair style is not an official part of LAMMPS - it exists in a separate fork. This is fairly common.

(2) It's possible that KOKKOS will automatically detect your architecture and you can delete that flag entirely. Unfortunately, it's hard to give good advice without access to your system.

**Reply by liubaoshuai1402** [reply]
Hi, @wcwitt 
I have installed lammps with mace.
But I counter some trouble when runing lammps.
Forgiving my poor knowledge about computers, I have to disturb you.
The main errors at log of lammps is at here:

terminate called after throwing an instance of 'c10::Error'
  what():  The specified pointer resides on host memory and is not registered with any CUDA device.
Exception raised from getDeviceFromPtr at ../aten/src/ATen/cuda/CUDADevice.h:16 (most recent call first):

a relatively complete version is:
<img width="886" alt="error" src="https://github.com/user-attachments/assets/a26b1299-4417-410e-a35c-6635c4907f9c">

I run lammps by "lmp -k on g 1 -sf kk -in in.lmp" using command line directly.
I wonder whether it is caused by my wrong configuration when compiling?
But i don not get a error from cmake.

One possible reason i guess:
I don't do the similar thing like "sintr -A YOUR-ACCOUNT-GPU -p ampere -N 1 --gres=gpu:1 -t 1:00:00"
I don't know how realize the similar effect by my own computer command.
I searched the meaning of sintr but get nothing.

Thanks!

**Reply by MicPellegrino** [reply]
Hi, I get the same error when running any single-precision model.
I initially thought it was an instability issue, but I am now convinced it must have to do either with Kokkos or with the script to convert to `.pt`.
I can run in ASE with the same time-step and in single-precision just fine, but I get the CUDA exception in LAMMPS even when reducting the time-step to 1/10th.

**Reply by MicPellegrino** [reply]
P.s. did you get any warning when converting to `.pt`?
I get:
```
UserWarning: Environment variable TORCH_FORCE_NO_WEIGHTS_ONLY_LOAD detected, since the`weights_only` argument was not explicitly passed to `torch.load`, forcing weights_only=False.
```
I think the exception may have something to do with how the model is loaded. I also think it should be loaded with `weights_only=True` (?), but for some reason that gets overwritten when setting the env. variable.
When passing `weights_only=True` as input to `torch.load()` in `create_lammps_model.py` I get:
```
UnpicklingError: Weights only load failed. This file can still be loaded, to do so you have two options, do those steps only if you trust the source of the checkpoint. 
	(1) In PyTorch 2.6, we changed the default value of the `weights_only` argument in `torch.load` from `False` to `True`. Re-running `torch.load` with `weights_only` set to `False` will likely succeed, but it can result in arbitrary code execution. Do it only if you got the file from a trusted source.
	(2) Alternatively, to load with `weights_only=True` please check the recommended steps in the following error message.
	WeightsUnpickler error: Unsupported global: GLOBAL mace.modules.models.ScaleShiftMACE was not an allowed global by default. Please use `torch.serialization.add_safe_globals([ScaleShiftMACE])` or the `torch.serialization.safe_globals([ScaleShiftMACE])` context manager to allowlist this global if you trust this class/function.

Check the documentation of torch.load to learn more about types accepted by default with weights_only https://pytorch.org/docs/stable/generated/torch.load.html.
```

**Reply by MicPellegrino** [reply]
(Or it could just be that

> Currently, there are no precision options with the KOKKOS package. All compilation and computation is performed in double precision.
https://www.afs.enea.it/software/lammps/doc17/html/accelerate_kokkos.html

And thus there's no hope to run MACE in sp with Kokkos...)

**Reply by lianxiliang** [reply]
> Hi, I get the same error when running any single-precision model. I initially thought it was an instability issue, but I am now convinced it must have to do either with Kokkos or with the script to convert to `.pt`. I can run in ASE with the same time-step and in single-precision just fine, but I get the CUDA exception in LAMMPS even when reducting the time-step to 1/10th.

Hi, it has been some time since you posted the issue. Have you solved the problem because I encountered the same issue? If so, it would be great to have a solution.

**Reply by MicPellegrino** [reply]
Hi, no luck so far. I am pretty much convinced it's a limitation of the LAMMPS-Kokkos or Kokkos-MACE interfaces, and not a numerical stability issue. I would place my hopes in ASE, openMM and possibly in the new Gromacs NNP interface (which should be developed and maintained by the core development team, contrarly to the LAMMPS one...).

**Reply by ilyes319** [reply]
We have a new interface to lammps using the MLIAP interface, I will add a doc soon.

**Reply by lianxiliang** [reply]
Thanks a lot for the information. It is great to know a solution is coming soon.
